src/multirust-cli/self_update.rs

```diff
@@ -0,0 +1,969 @@
+//   last it will be deleted when the handle closes.
+//
+// This is the DELETE_ON_CLOSE method from
+// http://www.catch22.net/tuts/self-deleting-executables
+//
+// ... which doesn't actually work because Windows won't really
+// delete a FILE_FLAG_DELETE_ON_CLOSE process when it exits.
+//
+// .. augmented with this SO answer
+// http://stackoverflow.com/questions/10319526/understanding-a-self-deleting-program-in-c
+#[cfg(windows)]
+fn delete_multirust_and_cargo_home() -> Result<()> {
+    use rand;
+
+    // CARGO_HOME, hopefully empty except for bin/multirust.exe
+    let ref cargo_home = try!(get_cargo_home());
+    // The multirust.exe bin
+    let ref multirust_path = cargo_home.join(&format!("bin/multirust{}", EXE_SUFFIX));
+
+    // The directory containing CARGO_HOME
+    let work_path = cargo_home.parent().expect("CARGO_HOME doesn't have a parent?");
+
+    // Generate a unique name for the files we're about to move out
+    // of CARGO_HOME.
+    let numbah: u32 = rand::random();
+    let gc_exe = work_path.join(&format!("multirust-gc-{:x}.exe", numbah));
+
+    use winapi::{FILE_SHARE_DELETE, FILE_SHARE_READ,
+                 INVALID_HANDLE_VALUE, FILE_FLAG_DELETE_ON_CLOSE,
+                 DWORD, SECURITY_ATTRIBUTES, OPEN_EXISTING,
+                 GENERIC_READ};
+    use kernel32::{CreateFileW, CloseHandle};
+    use std::os::windows::ffi::OsStrExt;
+    use std::ptr;
+    use std::io;
+    use std::mem;
+
+    unsafe {
+        // Copy multirust (probably this process's exe) to the gc exe
+        try!(utils::copy_file(multirust_path, &gc_exe));
+
+        let mut gc_exe_win: Vec<_> = gc_exe.as_os_str().encode_wide().collect();
+        gc_exe_win.push(0);
+
+        // Open an inheritable handle to the gc exe marked
+        // FILE_FLAG_DELETE_ON_CLOSE. This will be inherited
+        // by subsequent processes.
+        let mut sa = mem::zeroed::<SECURITY_ATTRIBUTES>();
+        sa.nLength = mem::size_of::<SECURITY_ATTRIBUTES>() as DWORD;
+        sa.bInheritHandle = 1;
+
+        let gc_handle = CreateFileW(gc_exe_win.as_ptr(),
+                                    GENERIC_READ,
+                                    FILE_SHARE_READ | FILE_SHARE_DELETE,
+                                    &mut sa,
+                                    OPEN_EXISTING,
+                                    FILE_FLAG_DELETE_ON_CLOSE,
+                                    ptr::null_mut());
+
+        if gc_handle == INVALID_HANDLE_VALUE {
+            let err = io::Error::last_os_error();
+            return Err(Error::WindowsUninstallMadness(err));
+        }
+
+        defer!{{ let _ = CloseHandle(gc_handle); }}
+
+        try!(Command::new(gc_exe).spawn()
+             .map_err(|e| Error::WindowsUninstallMadness(e)));
+        
+        // The catch 22 article says we must sleep here to give
+        // Windows a chance to bump the processes file reference
+        // count. acrichto though is in disbelief and *demanded* that
```

Comment:
I'd be fine if you want to leave my github username here as well so I can be maximally harassed